scripts/files/fs.py

```diff
@@ -42,3 +49,50 @@ def exists(path: str) -> bool:
     if is_s3(path):
         return fs_s3.exists(path)
     return fs_local.exists(path)
+
+
+def write_all(inputs: List[str], target: str, concurrency: Optional[int] = 10) -> List[str]:
+    """Writes list of files to target destination using multithreading.
+
+    Args:
+        inputs: list of files to read
+        target: target folder to write to
+
+    Returns:
+        list of written file paths
+    """
+    written_tiffs: List[str] = []
+    with ThreadPoolExecutor(max_workers=concurrency) as executor:
+        futuress = {
+            executor.submit(write, os.path.join(target, f"{input.split('/')[-1]}"), read(input)): input for input in inputs
+        }
+        for future in as_completed(futuress):
+            if future.exception():
+                get_log().warn("Failed Read-Write", error=future.exception())
+            else:
+                written_tiffs.append(future.result())
+
+    if len(inputs) != len(written_tiffs):
+        get_log().error("Missing Files", count=len(inputs) - len(written_tiffs))
+        raise Exception("Not all source files were written")
+    return written_tiffs
+
+
+def find_sidecars(inputs: List[str], extensions: List[str]) -> List[str]:
+    """Searches for sidecar files.
+     A sidecar files is a file with the same name as the input file but with a different extension.
+
+    Args:
+        inputs: list of input files to search for extensions
+        extensions: the sidecar file extensions
+
+    Returns:
+        list of existing sidecar files
+    """
+    sidecars = []
+    for file in inputs:
+        for extension in extensions:
+            sidecar = f"{file.split('.')[0]}{extension}"
```

Comment:
You'll get into an issue if `file` value contains multiple `.`. Using os.path is safer when dealing with paths:
```python
>>> import os
>>> file = "my.file.txt"
>>> f"{file.split('.')[0]}.doc"
'my.doc'
>>> file_prefix, ext = os.path.splitext(file)
>>> f"{file_prefix}.doc"
'my.file.doc'
```

Reply:
Awesome, thanks was looking into this too, a case of only testing with s3 paths beforehand :facepalm: 
I noticed exists was also quite slow and detracted from the speed increase of downloading the files so I made this function multithreading too.

Reply:
Do we have some runtimes to compare?

- with/without multithreading download
- same with sidecar files

Reply:
Here are some results for downloading the sidecars (average of 10 runs):
|                  | Before Changes                  | Multithread 1           | Multithread 25           |
|------------------|---------------------------------|-------------------------|--------------------------|
| Sidecar Download | 48s (download without checking) | 1min 23s (using exists) | 33s (multithread exists) |

I found it hard to test write_all but I re-ran waikato DEM and the [workflow using multithreading](https://argo.linzaccess.com/workflows/argo/test-mdavidson-multi-hb88v?tab=summary) took a total of 15mins rather than the [original without multithreading](https://argo.linzaccess.com/workflows/argo/waikato-2021-dem-1m2l7vx?tab=workflow) wich took 31 minutes.

My concern is that it still has the same `could not connect to endpoint url` error maybe somewhat more so with multithreading?